kia-lib/src/main/kotlin/com/faforever/ice/telemetry/TelemetryClient.kt

```diff
@@ -51,33 +54,47 @@ class TelemetryClient(
     init {
         logger.info {
             "Open the telemetry ui via ${
-                serverBaseUrl.replaceFirst(
-                    "ws",
-                    "http",
-                )
+                serverBaseUrl.replaceFirst("ws", "http")
             }/app.html?gameId=${iceOptions.gameId}&playerId=${iceOptions.userId}"
         }
 
         val uri: URI = URI.create("$serverBaseUrl/adapter/v1/game/${iceOptions.gameId}/player/${iceOptions.userId}")
         websocketClient = TelemetryWebsocketClient(uri)
 
-        connectingFuture = connectAsync()
+        sendingLoopThread = Thread(this::sendingLoop)
+        sendingLoopThread.uncaughtExceptionHandler = Thread.UncaughtExceptionHandler { _: Thread?, e: Throwable? ->
+            logger.error { "Thread sendingLoop crashed unexpectedly: $e" }
+        }
+
+        CompletableFuture.runAsync {
+            websocketClient.connectBlocking()
+            sendingLoopThread.start()
+        }
+
         registerAsPeer()
     }
 
-    private fun connectAsync() = CompletableFuture.runAsync(websocketClient::connect)
-        .exceptionally {
-            logger.error(it) { "Failed to connect to telemetry websocket" }
-            null
-        }
+    private fun sendingLoop() {
+        while (true) {
+            val message = messageQueue.take()
+
+            while (websocketClient.isClosed) {
+                // Before reconnecting delay 2, 8, 32, 64, 64, 64... seconds.
+                val delay = 2.shl((2 * connectionRetryAttempts).coerceAtMost(5))
```

Comment:
I'm not sure if this retry logic makes sense. Maybe we just want to try a couple of times and then give up rather than keep retrying.

Reply:
Yes. Also let's not reinvent the wheel, for these retry and catch failure problems we already have the [failsafe.dev](failsafe.dev) library on board.

It should support blocking too, [here](https://github.com/FAForever/kotlin-ice-adapter/blob/f3fcf74a87ff6d5ab5f4227c72390ede41d499d8/kia-lib/src/main/kotlin/com/faforever/ice/ice4j/AgentWrapper.kt#L95) is an async example: